aws/stackstate_checks/aws/aws.py

```diff
@@ -0,0 +1,290 @@
+            # TODO: finish creating S3 ARN
+            # service = 's3'
+            pass
+        elif resource_type == 'AWS::RDS::DBInstance':
+            # TODO: finish creating RDS ARN
+            # service = 'rds'
+            pass
+        elif resource_type == 'AWS::SNS::Topic':
+            # TODO: finish creating SNS ARN
+            # service = 'sns'
+            pass
+        elif resource_type == 'AWS::SQS::Queue':
+            # TODO: finish creating SQS ARN
+            # service = 'sqs'
+            pass
+        elif resource_type in ['AWS::DynamoDB::Table', 'AWS::DynamoDB', 'DynamoDB']:
+            arn_format = 'arn:aws:dynamodb:{0}:{1}:table/{2}'
+            try:
+                resource = segment['aws']['table_name']
+            except KeyError:
+                arn = segment['aws']['operation']
+        elif resource_type == 'AWS::EC2::Instance':
+            arn_format = 'arn:aws:ec2:{0}:{1}:instance/{2}'
+            try:
+                resource = segment['aws']['ec2']['instance_id']
+            except KeyError:
+                pass
+        else:
+            if segment.get('namespace') == 'local':
+                try:
+                    arn = self.arns[parent_span_id]
+                    self.arns[span_id] = arn
+                except KeyError:
+                    pass
+
+        if resource:
+            arn = arn_format.format(self.region, self.account_id, resource)
+
+        if arn:
+            self.arns[span_id] = arn
+
+        return arn
+
+    @staticmethod
+    def _send_payload(traces):
+        """Sends traces payload to Traces Agent."""
+        headers = {'Content-Type': 'application/json'}
+        requests.put(TRACES_API_ENDPOINT, data=json.dumps(traces), headers=headers)
+
+
+class AwsClient:
+    def __init__(self, instance):
+        aws_access_key_id = instance.get('aws_access_key_id')
+        aws_secret_access_key = instance.get('aws_secret_access_key')
+        role_arn = instance.get('role_arn')
+        self.region = instance.get('region')
+        self.aws_session_token = None
+
+        if aws_secret_access_key and aws_access_key_id and role_arn and self.region:
+            sts_client = boto3.client('sts', config=DEFAULT_BOTO3_CONFIG, aws_access_key_id=aws_access_key_id,
+                                      aws_secret_access_key=aws_secret_access_key)
+            role = sts_client.assume_role(RoleArn=role_arn, RoleSessionName='sts-agent-check')
+            self.aws_access_key_id = role['Credentials']['AccessKeyId']
+            self.aws_secret_access_key = role['Credentials']['SecretAccessKey']
+            self.aws_session_token = role['Credentials']['SessionToken']
+
+    def get_account_id(self):
+        return self._get_boto3_client('sts').get_caller_identity().get('Account')
+
+    def get_xray_traces(self):
+        xray_client = self._get_boto3_client('xray')
+
+        start_time = datetime.datetime.utcnow() - datetime.timedelta(minutes=1)
```

Comment:
We might have to look at this a bit. Specifically this part: `datetime.timedelta(minutes=1)`. If the interval of the check is changed we might lose / duplicate data. For now this is probably ok, but can you make a follow up ticket to introduce the functionality to take the check execution time into account

Reply:
How hard would it be for you to implement this functionality? Otherwise we can also just do it now

Reply:
We can fix this that we record end time (prehaps in file) so in we can have something like:
`if old_end_time is None:`
`....old_end_time = datetime.datetime.utcnow() - datetime.timedelta(minutes=1)`
`start_time = old_end_time`
`end_time = datetime.datetime.utcnow()`

Reply:
Yup should work. Let's add it to the check configuration as a optional value and log when we update this `old_end_time` variable so if the checks fails you can restart it at that point

Reply:
Added caching old end time in file. If file is not defined in YAML config its created in temporary directory.

First iteration wrote json cache file, but differences in Python 2.7 and 3 made conversion of time format from string to datetime and back pain. So I used pickle, its in binary format, but no conversion or timezone voodoo is needed as it (de)serializes pure python objects.

Reply:
I implemented cache that is not in binary format as Tiziano requested.